docs/articles/improve-raptor-with-rag.md

```diff
@@ -0,0 +1,561 @@
+            })
+    
+    main_df = pd.DataFrame(main_list)
+    unique_clusters = main_df["clusters"].unique()
+    if len(unique_clusters) == 0:
+        return df_clusters, pd.DataFrame(columns=["summaries", "level", "clusters"])
+
+    print(f"--Generated {len(unique_clusters)} clusters--")
+
+    summaries = []
+    for cluster in unique_clusters:
+        text_in_df = main_df[main_df["clusters"] == cluster]
+        unique_texts = text_in_df["text"].tolist()
+        text = "------\n------".join(unique_texts)
+        summary = generate_summary(text)
+        summaries.append(summary)
+
+    df_summaries = pd.DataFrame({
+        "summaries": summaries,
+        "level": [level] * len(summaries),
+        "clusters": unique_clusters
+    })
+
+    return df_clusters, df_summaries
+
+def recursive_embedding_with_cluster_summarization(
+    texts: List[str],
+    number_of_levels: int = 3,
+    level: int = 1,
+    target_dim: int = 10,
+    threshold: float = 0.1
+) -> Dict[int, Tuple[pd.DataFrame, pd.DataFrame]]:
+    if level > number_of_levels:
+        return {}
+    
+    results = {}
+    df_clusters, df_summaries = embed_cluster_summaries(texts, level, target_dim, threshold)
+    results[level] = (df_clusters, df_summaries)
+    
+    if df_summaries.empty or len(df_summaries['clusters'].unique()) == 1:
+        print(f"No more unique clusters found at level {level}. Stopping recursion.")
+        return results
+    
+    if level < number_of_levels:
+        next_level_texts = df_summaries['summaries'].tolist()
+        next_level_results = recursive_embedding_with_cluster_summarization(
+            next_level_texts, 
+            number_of_levels, 
+            level + 1,
+            target_dim,
+            threshold
+        )
+        results.update(next_level_results)
+    
+    return results
+ 
+```
+
+Okay, the code might seem a bit daunting at first glance, but don’t worry! Just give it a couple of looks, and it will start to make sense. Essentially, we’re just following the flow I mentioned earlier. 
+
+```python
+def process_text_hierarchy(
+    texts: List[str], 
+    number_of_levels: int = 3,
+    target_dim: int = 10,
+    threshold: float = 0.1
+) -> Dict[str, pd.DataFrame]:
+    hierarchy_results = recursive_embedding_with_cluster_summarization(
+        texts, number_of_levels, target_dim=target_dim, threshold=threshold
+    )
+    
+    processed_results = {}
+    for level, (df_clusters, df_summaries) in hierarchy_results.items():
+        if df_clusters.empty or df_summaries.empty:
+            print(f"No data for level {level}. Skipping.")
+            continue
+        processed_results[f"level_{level}_clusters"] = df_clusters
+        processed_results[f"level_{level}_summaries"] = df_summaries
+    
+    return processed_results
+
+results = process_text_hierarchy(chunks, number_of_levels=3)
+```
+
+![image.png](../assets/use_cases/improve-raptor-with-rag/raptor-3.png)
+
+### Inference
+
```

Comment:
pngpnfSo... the raptor tree and nodes etc discussed above (lines 403-5) aren't really a part of this diagram.. typically the text and diagram following it are supposed to reflect each other. This diagram is hard to read - text is small with a lot of white space. Does this look like a tree? If so, how?
There are some options here: 1 revise diagram to reflect RAPTOR tree description above (maybe something more like the attached img from the RAPTOR paper), 2 omit diagram.

Reply:
I think at this stage, I would accept the advise for omitting the diagram! 

Thanks